### PR TITLE
Enable the latest managed pgo data

### PR DESCRIPTION
**Pull request by davidwrighton**
- Implementation is parallel to existing ibc handling, so that it can be toggled on/off by adjusting the `UsingToolIbcOptimization` property
- Use the same data for all assemblies produced in current build
- Apply data to release builds only
- Disable mismatch assertions in jit for current state where il mismatches are common

### CODE REVIEW

**Comment by davidwrighton** [comment]
@dotnet/crossgen-contrib 
@agocke

**Review by AndyAyersMS — APPROVED** [review]
Jit changes ok (per offline discussion).

This data flows through to jitting (right?). If so we might want to look at what happens to jit codegen. I can do that separately though.

**Comment by mangod9** [comment]
Since IBC is currently not used/supported should that code be removed?

**Review by trylek — APPROVED** [review]
LGTM, thank you! Just for my curiosity, do the MIBC data restoration failures mean we first need to somehow seed the optimization data or are we going to need to make the restoration optional?

File: eng/restore/optimizationData.targets [hunk 1]
```diff
@@ -34,4 +34,50 @@
 
   </Target>
 
+
+  <ItemGroup>
+    <!-- Mibc data to use when exact architecture match is available -->
+    <MIBCPackageDef Include="optimization.windows_nt-x86.MIBC.Runtime" Version="$(optimizationwindows_ntx64MIBCRuntimeVersion)" MibcArchitecture="Windows/x86"/>
```
Comment (Lxiamail) [inline eng/restore/optimizationData.targets:40]: Should "optimizationwindows_ntx64MIBCRuntimeVersion" be "optimizationwindows_ntx86MIBCRuntimeVersion" instead in line 40?

File: eng/Versions.props [hunk 2]
```diff
@@ -116,6 +116,9 @@
     <SystemSecurityCryptographyX509CertificatesTestDataVersion>5.0.0-beta.21159.1</SystemSecurityCryptographyX509CertificatesTestDataVersion>
     <SystemWindowsExtensionsTestDataVersion>5.0.0-beta.21159.1</SystemWindowsExtensionsTestDataVersion>
     <!-- dotnet-optimization dependencies -->
+    <optimizationwindows_ntx64MIBCRuntimeVersion>99.99.99-master-20210317.2</optimizationwindows_ntx64MIBCRuntimeVersion>
+    <optimizationwindows_ntx86MIBCRuntimeVersion>99.99.99-master-20210317.2</optimizationwindows_ntx86MIBCRuntimeVersion>
+    <optimizationlinuxx64MIBCRuntimeVersion>99.99.99-master-20210317.2</optimizationlinuxx64MIBCRuntimeVersion>
```
Comment (Lxiamail) [inline eng/Versions.props:121]: Is nuget package "optimization.MIBC.Runtime.99.99.99-master-20210317.2.nupkg" needed anywhere?

File: eng/Versions.props [hunk 3]
```diff
@@ -116,6 +116,9 @@
     <SystemSecurityCryptographyX509CertificatesTestDataVersion>5.0.0-beta.21159.1</SystemSecurityCryptographyX509CertificatesTestDataVersion>
     <SystemWindowsExtensionsTestDataVersion>5.0.0-beta.21159.1</SystemWindowsExtensionsTestDataVersion>
     <!-- dotnet-optimization dependencies -->
+    <optimizationwindows_ntx64MIBCRuntimeVersion>99.99.99-master-20210317.2</optimizationwindows_ntx64MIBCRuntimeVersion>
```
Comment (Lxiamail) [inline eng/Versions.props:119]: I heard that there is a way to specify version using wildcard, like in versions.prop "<optimizationlinuxx64MIBCRuntimeVersion>99.99.99-master-20210317.2</optimizationlinuxx64MIBCRuntimeVersion>". instruction doc is https://docs.microsoft.com/en-us/nuget/concepts/package-versioning. However, I haven't tried it before. I see using wildcard maybe helpful if later on before preview3 release that we need to update training packages again for some reason and we don't need to updated version in runtime repo. However, I think the risk is quite low as long as you have tested the current training packages.

Comment (davidwrighton) [reply eng/Versions.props:119]: We don't want wildcards here. If we use exact versions, we can tie into the dependency flow.

Comment (Lxiamail) [reply eng/Versions.props:119]: Yes, we whan tie to dependency flow.

File: eng/Versions.props (hunk 2 above)
Comment (davidwrighton) [reply eng/Versions.props:121]: Ideally the complex logic that is used to choose which one of these applies to a given platform will change to use the non-arch specific nuget package, but for now I just made it work without it.

File: eng/restore/optimizationData.targets (hunk 1 above)
Comment (davidwrighton) [reply eng/restore/optimizationData.targets]: Yes. Fixed.

**Comment by AndyAyersMS** [comment]
@DrewScoggins @adamsitnik 

FYI, this PR may lead to widespread changes in benchmark perf as static PGO data will now be available to the jit for some assemblies. Hopefully, many many improvements...

**Pull request merged**